airbyte-integrations/connectors/source-github/source_github/streams.py

```diff
@@ -293,85 +354,122 @@ class PullRequests(SemiIncrementalGithubStream):
         "requested_reviewers",
         "requested_teams",
     )
-    stream_base_params = {
-        "state": "all",
-        "sort": "updated",
-        "direction": "desc",
-    }
 
-    def read_records(self, **kwargs) -> Iterable[Mapping[str, Any]]:
-        with vcr.use_cassette(cache_file.name, record_mode="new_episodes", serializer="json"):
-            yield from super().read_records(**kwargs)
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self._first_read = True
+
+    # TODO Fix vcr error:
+    #   UnicodeDecodeError: 'utf-8' codec can't decode byte 0x72 in position 1: invalid start byteDoes this HTTP
+    #   interaction contain binary data? If so, use a different serializer (like the yaml serializer) for this request?
+    # def read_records(self, **kwargs) -> Iterable[Mapping[str, Any]]:
+    #     with vcr.use_cassette(cache_file.name, record_mode="new_episodes", serializer="json"):
+    #         yield from super().read_records(**kwargs)
+
+    def read_records(self, stream_state: Mapping[str, Any] = None, **kwargs) -> Iterable[Mapping[str, Any]]:
+        """
+        Decide if this a first read or not by the presence of the state object
+        """
+        self._first_read = not bool(stream_state)
+        yield from super().read_records(stream_state=stream_state, **kwargs)
 
     def path(self, **kwargs) -> str:
         return f"repos/{self.repository}/pulls"
 
-    def transform(self, record: Mapping[str, Any]) -> Mapping[str, Any]:
+    def transform(self, record: MutableMapping[str, Any]) -> MutableMapping[str, Any]:
         record = super().transform(record=record)
 
-        head = record.get("head", {})
-        head_user = head.pop("user", None)
-        head["user_id"] = head_user.get("id") if head_user else None
-        head_repo = head.pop("repo", None)
-        head["repo_id"] = head_repo.get("id") if head_repo else None
-
-        base = record.get("base", {})
-        base_user = base.pop("user", None)
-        base["user_id"] = base_user.get("id") if base_user else None
-        base_repo = base.pop("repo", None)
-        base["repo_id"] = base_repo.get("id") if base_repo else None
+        for nested in ("head", "base"):
+            entry = record.get(nested, {})
+            entry["user_id"] = record.get("head", {}).pop("user", {}).get("id")
+            entry["repo_id"] = record.get("head", {}).pop("repo", {}).get("id")
 
         return record
 
+    def request_params(self, **kwargs) -> MutableMapping[str, Any]:
+        base_params = super().request_params(**kwargs)
+        params = {"state": "all", "sort": "updated", "direction": "desc" if self.is_sorted_descending else "asc"}
```

Comment:
```suggestion
        # The very first time we read this stream we want to read ascending so we can save state in case of 
        # a halfway failure. But if there is state, we read descending to allow incremental behavior. 
        params = {"state": "all", "sort": "updated", "direction": "desc" if self.is_sorted_descending else "asc"}
```